File: src/anylayout/extras/RelativeConstraints.java
package anylayout.extras;

import anylayout.Constraint;
import anylayout.LayoutContext;
import anylayout.LayoutInfo;
import static anylayout.extras.ConstraintBuilder.buildConstraint;
import fj.F;

import java.awt.Component;

public final class RelativeConstraints
{
	public static Constraint above(final Component below,final int padding)
	{
		return buildConstraintSameLeftAs(below).setTop(new F<LayoutContext, Integer>()
		{
			public Integer f(final LayoutContext context)
			{
				return context.getLayoutInfo(below).getOffset()-padding-context.getPreferredSize();
			}
		}).setWidth(ConstraintBuilder.preferredSize).setHeight(ConstraintBuilder.preferredSize);
	}

	private static ConstraintBuilderStage2 buildConstraintSameLeftAs(final Component component)
	{
		return buildConstraint().setLeft(equalTo(component));
	}

	public static Constraint below(final Component above,final int padding)
	{
		return buildConstraintSameLeftAs(above).setTop(after(above, padding)).setWidth(ConstraintBuilder.preferredSize).setHeight(ConstraintBuilder.preferredSize);
	}

	private static F<LayoutContext,Integer> after(final Component above,final int padding)
	{
		return new F<LayoutContext,Integer>()
		{
			public Integer f(final LayoutContext context)
			{
				return context.getLayoutInfo(above).getFarOffset()+padding;
			}
		};
	}

	public static Constraint rightOf(final Component left,final int padding)
	{
		return buildConstraint().setLeft(after(left, padding)).setTop(equalTo(left)).setWidth(ConstraintBuilder.preferredSize).setHeight(ConstraintBuilder.preferredSize);
	}

	public static Constraint halfwayBetween(final Component one,final Component two)
	{
		return buildConstraint().setLeft(new F<LayoutContext, Integer>()
		{
			public Integer f(final LayoutContext context)
			{
				final LayoutInfo oneInfo=context.getLayoutInfo(one);
				final LayoutInfo twoInfo=context.getLayoutInfo(two);

				return Math.max(0, (oneInfo.getFarOffset()+twoInfo.getOffset()-context.getPreferredSize())/2);
			}
		}).setTop(new F<LayoutContext, Integer>()
		{
			public Integer f(final LayoutContext context)
			{
				return (after(one, context.getLayoutInfo(two).getOffset()).f(context)-context.getPreferredSize())/2;
			}
		}).setWidth(ConstraintBuilder.preferredSize).setHeight(ConstraintBuilder.preferredSize);
	}

	public static Constraint levelWith(final Component horizontal,final Component vertical)
	{
		return buildConstraintSameLeftAs(vertical).setTop(equalTo(horizontal)).setWidth(ConstraintBuilder.preferredSize).setHeight(ConstraintBuilder.preferredSize);
	}

	private static F<LayoutContext,Integer> equalTo(final Component horizontal)
	{
		return new F<LayoutContext,Integer>()
		{
			public Integer f(final LayoutContext context)
			{
				return context.getLayoutInfo(horizontal).getOffset();
			}
		};
	}
}

File: src/anylayout/extras/RelativeConstraints.scala
package anylayout.extras

import anylayout._
import ConstraintBuilder.buildConstraint

import java.awt.Component

object RelativeConstraints {
 def above(below: Component, padding: Int) = buildConstraintSameLeftAs(below) setTop (context => context.layoutInfoFor(below).getOffset.intValue-padding-context.preferredSize) setWidth ConstraintBuilder.preferredSize setHeight ConstraintBuilder.preferredSize

 def buildConstraintSameLeftAs(component: Component) = buildConstraint setLeft equalTo(component)
 def below(above: Component, padding: Int) = buildConstraintSameLeftAs(above) setTop after(above, padding) setWidth ConstraintBuilder.preferredSize setHeight ConstraintBuilder.preferredSize
 def after(above: Component, padding: Int): LayoutContext => Int = _.layoutInfoFor(above).getFarOffset.intValue+padding
 def rightOf(left: Component, padding: Int) = buildConstraint setLeft after(left, padding) setTop equalTo(left) setWidth ConstraintBuilder.preferredSize setHeight ConstraintBuilder.preferredSize
 def halfwayBetween(one: Component, two: Component) = buildConstraint setLeft (c => {
  val oneInfo = c.layoutInfoFor(one)
  val twoInfo = c.layoutInfoFor(two)
  val tmp = (oneInfo.getFarOffset.intValue + twoInfo.getOffset.intValue - c.preferredSize) / 2
  Math.max(0, tmp / 2) }
                                                                              ) setTop (context => (after(one, context.layoutInfoFor(two).getOffset.intValue)(context).intValue-context.preferredSize.intValue)/2) setWidth ConstraintBuilder.preferredSize setHeight ConstraintBuilder.preferredSize

 def levelWith(horizontal: Component, vertical: Component) = buildConstraintSameLeftAs(vertical).setTop(equalTo(horizontal)).setWidth(ConstraintBuilder.preferredSize).setHeight(ConstraintBuilder.preferredSize)
 private def equalTo(horizontal: Component): LayoutContext => Int = _.layoutInfoFor(horizontal).getOffset
}
